gleambatch.py

```diff
@@ -0,0 +1,227 @@
+    FTData.propagate_down(rs)
+
+    # Propagate estimates upwards to super-regions
+    rs.fix_min_est(
+        "est_active", keep_nones=True, minimum_from="csse_active", minimum_mult=2.0
+    )  ## TODO: param for mult
+
+    rs.check_missing_estimates("est_active")
+
+    # Store the initial estimates in batch data
+    rs.fix_min_est("est_active")
+    batch.store_region_estimates(rs, "est_active", "FT_Infected")
+    # Finally, propagate csse upwards and also store
+    for loc_key, rem_key in [
+        ("csse_deaths", "JH_Deaths"),
+        ("csse_confirmed", "JH_Confirmed"),
+        ("csse_recovered", "JH_Recovered"),
+        ("csse_active", "JH_Infected"),
+    ]:
+        # Fix one-city provinces
+        for r in rs.regions:
+            if (
+                r.parent is not None
+                and len(r.parent.sub) == 1
+                and r.est.get(loc_key) is None
+            ):
+                r.est[loc_key] = r.parent.est.get(loc_key)
+        # Propagate up
+        rs.fix_min_est(loc_key)
+        # Store in batch
+        batch.store_region_estimates(rs, loc_key, rem_key)
+
+
+def estimates_to_gleamdef(batch, rs: Regions, input_xml_path):
+    gv = GleamDef(input_xml_path)
+    gv.set_start_date(batch.config["start_date"])
+    gv.clear_seeds()
+    for comp, coef in batch.config["compartments_mult"].items():
+        gv.add_seeds(rs, est_key="est_active", compartments={comp: coef})
+
+    return gv
+
+
+def parameterize(batch, gv):
+    for mit in batch.config["mitigations"]:
+        for sce in batch.config["scenarios"]:
+            gv2 = gv.copy()
+            gv2.set_seasonality(sce["param_seasonalityAlphaMin"])
+            gv2.set_traffic_occupancy(sce["param_occupancyRate"])
+            gv2.set_beta(mit["param_beta"])
+            gv2id = "{}.574".format(random.randint(1700000000000, 1800000000000))
+            gv2.set_id(gv2id)
+            gv2.set_name(gv2.full_name(batch.name))
+            sim = Simulation(gv2, None)
+            batch.add_simulation_info(
+                sim,
+                name=sce["name"],
+                group=mit["label"],
+                color=sce.get("color"),
+                style=sce.get("style"),
+            )
+
+
+def generate(args):
+    "Run the 'generate' subcommand"
+
+    with open(args.CONFIG_YAML, "rt") as f:
+        config = yaml.load(f)
+    Path(config["output_dir"]).expanduser().mkdir(exist_ok=True, parents=True)
+    batch = Batch.new(config, suffix=args.comment.replace(" ", "-"))
+
+    log.info(f"Reading regions from {batch.config['regions_file']} ...")
+    rs = Regions.load_from_yaml(batch.config["regions_file"])
+
+    estimate(batch, rs)
+
+    gv = estimates_to_gleamdef(batch, rs, args.GLEAM_XML)
+
+    parameterize(batch, gv)
+
+    batch.save_sim_defs_to_gleam()
+
+    batch.save()
+
+    log.info(
+        f"Run '{sys.argv[0]} process {batch.get_batch_file_path()}' after"
```

Comment:
It's using a global variable - is that really necessary? Would pass it there as a regular argument if it's needed (but IMO not that much here).

Reply:
You mean `log`, `sys.argv` or `batch` (not global)?

Reply:
I meant `sys.argv` here (log is "allowed" in my head)